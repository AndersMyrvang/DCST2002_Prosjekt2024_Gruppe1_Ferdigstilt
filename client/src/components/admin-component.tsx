import React, { useState, useEffect } from 'react';
import { User } from '../types';

const AdminPanel = () => {
  const [users, setUsers] = useState<User[]>([]);

  const fetchUsers = () => {
    fetch('/api/users')
      .then((response) => response.json())
      .then((data) => setUsers(data));
  };

  useEffect(() => {
    fetchUsers();

    const interval = setInterval(fetchUsers, 60000);

    return () => clearInterval(interval);
  }, []);

  // Sjekker om en bruker er online basert på last_login og last_logout
  const isUserOnline = (user: User) => {
    if (
      user.last_login &&
      (!user.last_logout || new Date(user.last_login) > new Date(user.last_logout))
    ) {
      return true; 
    }
    return false;
  };

  // Funksjon for å endre admin-status
  const toggleAdminStatus = (userId: number, isAdmin: boolean) => {
    fetch(`/api/users/${userId}/set-admin`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ isAdmin: !isAdmin }),
    })
      .then((response) => {
        if (response.ok) {
          setUsers(
            users.map((user) => (user.user_id === userId ? { ...user, is_admin: !isAdmin } : user)),
          );
        } else {
          console.error('Feil ved oppdatering av admin-status');
        }
      })
      .catch((error) => console.error('Feil ved toggleAdminStatus:', error));
  };

  return (
    <div className="admin-panel">
      <table>
        <thead>
          <tr>
            <th>User ID</th>
            <th>Username</th>
            <th>Email</th>
            <th>Member Since</th>
            <th>Online Status</th>
            <th>Admin</th>
          </tr>
        </thead>
        <tbody>
          {users.map((user) => (
            <tr key={user.user_id}>
              <td>{user.user_id}</td>
              <td>{user.username}</td>
              <td>{user.email}</td>
              <td>{new Date(user.first_login).toLocaleDateString()}</td>
              <td>
                <span className={isUserOnline(user) ? 'online-dot' : 'offline-dot'}></span>
              </td>
              <td>
                <input
                  type="checkbox"
                  checked={user.is_admin}
                  onChange={() => toggleAdminStatus(user.user_id!, user.is_admin)}
                />
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
};

export default AdminPanel;
