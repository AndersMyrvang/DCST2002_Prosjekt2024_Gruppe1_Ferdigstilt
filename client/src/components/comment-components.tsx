import * as React from 'react';
import { Component } from 'react-simplified';
import { Alert, Card, Row, Column, Form, Button, Card2, Card3 } from '../widgets';
import { NavLink } from 'react-router-dom';
import commentService from '../services/comment-service';
import userService from '../services/user-service';
import { Comment, User } from '../types';
import { createHashHistory } from 'history';
import { AiFillLike, AiFillDislike } from 'react-icons/ai';
import { FaTrashAlt } from 'react-icons/fa';
import { BiSolidPencil } from 'react-icons/bi';

const history = createHashHistory();

interface CommentSectionProps {
  page_id: number;
  onNewComment: () => void;
}

//Komponent for ny kommentar
export class NewComment extends Component<CommentSectionProps> {
  comment: Comment = {
    comment_id: 0,
    user_id: 0,
    content: '',
    created_at: new Date(),
    updated_at: new Date(),
    page_id: 0,
    likes: 0,
  };

  render() {
    return (
      <>
        <Card title="New Comment">
          <Column>
            <Form.Input
              type="text"
              aria-label={`Comment input`}
              value={this.comment.content}
              onChange={(event) => (this.comment.content = event.currentTarget.value)}
              rows={3}
            />
          </Column>
          <Column>
            <Button.Success
              onClick={() => {
                this.NewComment(this.comment.user_id, this.comment.content, this.props.page_id);
                this.comment.content = '';
              }}
            >
              {' '}
              SEND{' '}
            </Button.Success>
          </Column>
        </Card>
      </>
    );
  }
  mounted() {
    const currentUserId = localStorage.getItem('user_id');
    if (currentUserId != null) this.comment.user_id = parseInt(currentUserId, 10);
  }
  NewComment(user_id: number, content: string, page_id: number) {
    commentService
      .create(user_id, content, page_id)
      .then((comment) => {
        this.props.onNewComment();
      })
      .catch((error) => {
        console.error('Error creating comment:', error);
      });
  }
}

//Komponent for kommentarseksjonen (visning av kommentarer p√• en spesifikk side)
export class CommentSection extends Component<CommentSectionProps> {
  comment: Comment[] = [];
  user: User = {
    user_id: 0,
    google_id: '',
    username: '',
    email: '',
    profile_image_url: '',
    first_login: new Date(),
    last_login: new Date(),
    last_logout: new Date(),
    is_admin: false,
  };
  updatingComment: number | null = null;
  updatedContent: string = '';

  render() {
    return (
      <>
        <Card title="Comments">
          {this.user.user_id != 0 ? (
            <NewComment page_id={this.props.page_id} onNewComment={() => this.mounted()} />
          ) : null}
          {this.comment.map((comment) => (
            <Row key={new Date(comment.created_at).getTime()}>
              <Column>
                {' '}
                <img
                  src={`${comment.profile_image_url}`}
                  alt={`Bilde fra ${comment.profile_image_url}`}
                  width={'45px'}
                  style={{
                    borderRadius: '50%',
                    objectFit: 'cover',
                  }}
                />
              </Column>
              <Column>
                <b>{comment.username}</b>
              </Column>
              <Column>
                {this.updatingComment == comment.comment_id ? (
                  <Form.Input
                    type="text"
                    value={this.updatedContent}
                    onChange={(event) => (this.updatedContent = event.currentTarget.value)}
                  />
                ) : (
                  comment.content
                )}
              </Column>

              <Column>
                {this.user.user_id == comment.user_id || this.user.is_admin == true ? (
                  <Button.Danger
                    onClick={() => {
                      this.deleteComment(comment.comment_id);
                      this.mounted();
                    }}
                  >
                    <FaTrashAlt />
                  </Button.Danger>
                ) : null}
              </Column>
              <Column>
                {this.user.user_id == comment.user_id || this.user.is_admin == true ? (
                  <>
                    {this.updatingComment == comment.comment_id ? (
                      <>
                        <Button.Success
                          onClick={() => {
                            this.updateComment(comment.comment_id, this.updatedContent);
                            this.updatingComment = null;
                            this.mounted();
                          }}
                        >
                          Lagre
                        </Button.Success>
                      </>
                    ) : (
                      <>
                        <Button.Light
                          onClick={() => {
                            this.updatingComment = comment.comment_id;
                            this.updatedContent = comment.content;
                          }}
                        >
                          <BiSolidPencil />
                        </Button.Light>
                      </>
                    )}
                  </>
                ) : null}
              </Column>
            </Row>
          ))}
        </Card>
      </>
    );
  }

  mounted() {
    const current_userId = localStorage.getItem('user_id');
    if (current_userId != null) {
      userService
        .getUserById(parseInt(current_userId, 10))
        .then((user: User) => {
          this.user = user;
        })
        .catch((error) => {
          console.error('Error getting user: ', error);
          Alert.danger('Error getting user: ' + error.message);
        });
    }

    commentService
      .listComments(this.props.page_id)
      .then(async (comments: Comment[]) => {
        this.comment = comments;
      })
      .catch((error) => {
        console.error('Error fetching comments:', error);
        Alert.danger('Error getting comments: ' + error.message);
      });
  }

  deleteComment(comment_id: number) {
    commentService.deleteComment(comment_id);
  }
  updateComment(comment_id: number, content: string) {
    commentService.update(comment_id, content);
  }
}

export default CommentSection;
