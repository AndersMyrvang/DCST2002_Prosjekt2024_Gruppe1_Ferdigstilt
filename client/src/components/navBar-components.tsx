import React, { useState, useRef, useEffect } from 'react';
import SearchBar from './searchBar-component';
import { FaHouse } from 'react-icons/fa6';
import { User } from '../types';

interface HamburgerProps {
  user: User | null;
}

function DarkModeButton() {
  const [darkMode, setDarkMode] = useState(() => {
    const savedMode = localStorage.getItem('darkMode');
    return savedMode === 'true';
  });

  const toggleDarkMode = () => {
    const newMode = !darkMode;
    setDarkMode(newMode);
    document.body.classList.toggle('dark-mode', newMode);
    localStorage.setItem('darkMode', newMode.toString());
  };

  useEffect(() => {
    if (darkMode) {
      document.body.classList.add('dark-mode');
    }
  }, [darkMode]);

  return (
    <button className="dark-mode-button" onClick={toggleDarkMode} aria-label="Toggle Dark Mode">
      <i
        className={darkMode ? 'fas fa-sun' : 'fas fa-moon'}
        style={{ fontSize: '24px', color: darkMode ? '#FFD700' : '#555' }}
      />
    </button>
  );
}

const Hamburger: React.FC<HamburgerProps> = ({ user }) => {
  const [burgerClass, setBurgerClass] = useState('burger-bar unclicked');
  const [menuClass, setMenuClass] = useState('menu hidden');
  const [isMenuClicked, setIsMenuClicked] = useState(false);

  const [isLoggedIn, setIsLoggedIn] = useState(() => localStorage.getItem('isLoggedIn') === 'true');

  const [userName, setUserName] = useState(localStorage.getItem('user') || '');
  const menuRef = useRef<HTMLDivElement>(null);

  const updateMenu = () => {
    setIsMenuClicked((prev) => !prev);
    setBurgerClass(isMenuClicked ? 'burger-bar unclicked' : 'burger-bar clicked');
    setMenuClass(isMenuClicked ? 'menu hidden' : 'menu visible');
  };

  const handleClickOutside = (event: MouseEvent) => {
    if (menuRef.current && !menuRef.current.contains(event.target as Node)) {
      closeMenu();
    }
  };

  const closeMenu = () => {
    setBurgerClass('burger-bar unclicked');
    setMenuClass('menu hidden');
    setIsMenuClicked(false);
  };

  const handleSearch = (query: string) => {
    fetch(`/api/search?q=${query}`)
      .then((response) => response.json())
      .then((data) => {})
      .catch((error) => {
        console.error('Feil ved søk:', error);
      });
  };

  const handleLogout = () => {
    fetch('/api/logout', { method: 'GET', credentials: 'include' })
      .then((response) => {
        if (response.ok) {
          setIsLoggedIn(false);
          setUserName('');
          localStorage.removeItem('isLoggedIn');
          localStorage.removeItem('user');
          localStorage.removeItem('user_id');
          window.location.href = '/';
        } else {
          console.error('Logout failed');
        }
      })
      .catch((error) => console.error('Logout error:', error));
  };

  useEffect(() => {
    document.addEventListener('mousedown', handleClickOutside);

    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, []);

  // Sjekker om bruker er pålogget
  useEffect(() => {
    fetch('/api/current_user')
      .then((response) => response.json())
      .then((data) => {
        if (data && data.username) {
          setIsLoggedIn(true);
          setUserName(data.username);
          localStorage.setItem('isLoggedIn', 'true');
          localStorage.setItem('user', data.username);
          localStorage.setItem('user_id', data.user_id);
        } else {
          setIsLoggedIn(false);
          localStorage.setItem('isLoggedIn', 'false');
        }
      })
      .catch(() => {
        setIsLoggedIn(false);
        localStorage.setItem('isLoggedIn', 'false');
      });
  }, []);

  return (
    <div style={{ width: '100%' }} ref={menuRef}>
      <nav>
        <div className="house">
          <a href="/">
            <FaHouse />
          </a>
        </div>
        <div className="burger-menu" onClick={updateMenu} aria-label="Toggle Menu">
          <div className={burgerClass}></div>
          <div className={burgerClass}></div>
          <div className={burgerClass}></div>
        </div>

        <SearchBar onSearch={handleSearch} />

        <DarkModeButton />
      </nav>

      <div className={menuClass}>
        <ul className="menu-links" onClick={closeMenu} role="menu">
          <li>
            <a href="/">Home</a>
          </li>
          <li>
            <a href="#players">Players</a>
          </li>
          <li>
            <a href="#teams">Teams</a>
          </li>
          <li>
            <a href="#leagues">Leagues</a>
          </li>
          <li>
            <a href="#tags">Tags</a>
          </li>

          {!isLoggedIn ? (
            <li>
              <a href="https://accounts.google.com/signup/v2/createaccount">Register </a> /
              <a href="/auth/google"> Log in with Google </a>
            </li>
          ) : (
            <>
              <li>
                <a href="#/profile">My profile </a> /{' '}
                <a href="#" onClick={handleLogout}>
                  Log out
                </a>
              </li>
            </>
          )}
        </ul>
      </div>
    </div>
  );
};

export default Hamburger;
