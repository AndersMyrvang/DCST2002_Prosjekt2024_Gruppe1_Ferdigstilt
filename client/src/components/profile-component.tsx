import React, { useEffect, useState } from 'react';
import AdminPanel from './admin-component';
import { UserProfile } from '../types';

export const useUser = () => {
  const [user, setUser] = useState<UserProfile | null>(null);

  useEffect(() => {
    fetch('/api/current_user')
      .then((response) => response.json())
      .then((data) => {
        if (data) {
          setUser({
            username: data.username,
            email: data.email,
            photo: data.photo,
            firstLogin: new Date(data.firstLogin).toLocaleDateString(),
            lastLogin: new Date(data.lastLogin).toLocaleDateString(),
            is_admin: Boolean(data.is_admin),
            user_id: data.user_id,
          });
        }
      })
      .catch((error) => console.error('Error fetching user data:', error));
  }, []);

  return user;
};

const Profile = () => {
  const user = useUser();
  if (!user) {
    return <div>Laster...</div>;
  }

  return (
    <div className="profile-page">
      <div className="profile-card">
        <div className="profile-card-content">
          <div className="profile-header">
            <img src={user.photo} alt="Profilbilde" className="profile-img" />
          </div>
          <div className="profile-content">
            <h2 className="profile-name">{user.username}</h2>
            <p className="profile-info">Member since {user.firstLogin}</p>
            <p className="profile-details">
              <span>Last visited {user.lastLogin}</span>
              <br />
              <br />
              <span>{user.email}</span>
            </p>
          </div>
        </div>
      </div>

      {user.is_admin && (
        <div className="admin-panel-container">
          <h2>Admin Panel</h2>
          <AdminPanel />
        </div>
      )}
    </div>
  );
};

export default Profile;
