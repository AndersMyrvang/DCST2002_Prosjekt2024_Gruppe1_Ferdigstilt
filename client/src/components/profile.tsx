import React, { useEffect, useState } from 'react';

// Info om bruker som skal vises
interface UserProfile {
  displayName: string;
  email: string;
  photo: string;
}

// Lager komponent
const Profile = () => {
  const [user, setUser] = useState<UserProfile | null>(null);

  useEffect(() => {
    fetch('/api/current_user')
      .then((response) => response.json())
      .then((data) => {
        if (data) {
          setUser({
            displayName: data.displayName,
            email: data.emails && data.emails[0]?.value,
            photo: data.photos && data.photos[0]?.value,
          });
        }
      });
  }, []);

  // Dersom det ikke finnes eller laster inn tregt
  if (!user) {
    return <div>Laster...</div>;
  }

  // Viser info
  return (
    <div style={{ textAlign: 'center', marginTop: '20px' }}>
      <img
        src={user.photo}
        alt="Profilbilde"
        style={{ borderRadius: '50%', width: '150px', height: '150px', marginBottom: '20px' }}
      />
      <h2>{user.displayName}</h2>
      <p>{user.email}</p>
      <h3>"Vise lister eller noe annet her?"</h3>
    </div>
  );
};

export default Profile;
