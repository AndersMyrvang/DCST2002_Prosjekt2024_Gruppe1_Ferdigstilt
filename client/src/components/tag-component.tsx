import { Component } from 'react-simplified';
import * as React from 'react';
import { Alert, Card, Row, Column, Form, Button, Card2, Card3 } from '../widgets';
import { FaTrashAlt } from 'react-icons/fa';
import { NavLink } from 'react-router-dom';
import { Comment, Tag, User } from '../types';
import { createHashHistory } from 'history';
import tagService from '../services/tag-service';
import userService from '../services/user-service';

const history = createHashHistory();

//Komponent som viser liste over alle tags
export class TagList extends Component {
  tags: Tag[] = [];
  tagPageCounts: { [tag_id: number]: number } = {};
  user: User = {
    user_id: 0,
    google_id: '',
    username: '',
    email: '',
    profile_image_url: '',
    first_login: new Date(),
    last_login: new Date(),
    is_admin: false,
  };

  deleteTag(tagId: number) {
    tagService
      .deleteTag(tagId)
      .then(() => {
        this.tags = this.tags.filter((tag) => tag.tag_id !== tagId);
        this.forceUpdate();
      })
      .catch((error) => {
        console.error('Error deleting tag:', error);
        Alert.danger('Error deleting tag: ' + error.message);
      });
  }

  render() {
    return (
      <>
        <Card title="Tags">
          {this.tags.map((tag) => (
            <Row key={tag.tag_id}>
              <div
                style={{
                  justifyContent: 'space-between',
                  alignItems: 'center',
                  display: 'flex',
                  width: '100%',
                }}
              >
                <Column width={10}>
                  <a>{tag.tag_name}</a>
                </Column>
                <Column width={1}>
                  <span>Pages: {this.tagPageCounts[tag.tag_id] || 0}</span>
                </Column>
                <Column width={1}>
                  <div style={{ textAlign: 'right', marginLeft: '3em' }}>
                    {this.user.is_admin == true ? (
                      <Button.Danger
                        onClick={() => {
                          if (
                            window.confirm(
                              `Are you sure you want to delete the tag "${tag.tag_name}"?`,
                            )
                          ) {
                            this.deleteTag(tag.tag_id);
                          }
                        }}
                      >
                        <FaTrashAlt />
                      </Button.Danger>
                    ) : null}
                  </div>
                </Column>
              </div>
            </Row>
          ))}
        </Card>
        <Row>
          <Column>
            {this.user.user_id != 0 ? (
              <Button.Success onClick={() => history.push('/tags/new')}>New tag</Button.Success>
            ) : null}
          </Column>
        </Row>
      </>
    );
  }

  mounted(): void {
    tagService
      .getTags()
      .then((tags: Tag[]) => {
        this.tags = tags;
        const tagPageCounts: { [tag_id: number]: number } = {};
        const promises = tags.map((tag) =>
          tagService.getTagPageCount(tag.tag_id).then((pageCount: number) => {
            tagPageCounts[tag.tag_id] = pageCount;
          }),
        );
        Promise.all(promises)
          .then(() => {
            this.tagPageCounts = tagPageCounts;
            this.forceUpdate();
          })
          .catch((error) => {
            console.error('Error fetching page counts:', error);
          });
      })
      .catch((error: any) => {
        console.error('Error fetching tags:', error);
        Alert.danger('Error getting tags: ' + error.message);
      });

    const current_userId = localStorage.getItem('user_id');
    if (current_userId != null) {
      userService
        .getUserById(parseInt(current_userId, 10))
        .then((user: User) => {
          this.user = user;
        })
        .catch((error) => {
          console.error('Error getting user: ', error);
          Alert.danger('Error getting user: ' + error.message);
        });
    }
  }
}

//Komponent for Ã¥ legge til tags
export class TagAdd extends Component {
  tag = { tag_id: 0, tag_name: '' };

  add = async () => {
    try {
      const result = await tagService.addTag(this.tag);

      Alert.success('Tag added');

      history.push('/tags');
    } catch (error) {
      console.error('Error adding tag:', error);
      if (error instanceof Error) {
        Alert.danger('Error adding tag: ' + error.message);
      } else {
        Alert.danger('Error adding tag');
      }
    }
  };

  render() {
    return (
      <Card title="Add tag">
        <Form.Label>Name</Form.Label>
        <Form.Input
          type="text"
          value={this.tag.tag_name}
          placeholder="Tag name"
          onChange={(event) => {
            this.tag.tag_name = event.currentTarget.value;
            this.forceUpdate();
          }}
        />
        <Button.Success onClick={this.add}>Add tag</Button.Success>
        <Button.Danger onClick={() => history.push('/tags')}>Cancel</Button.Danger>
      </Card>
    );
  }
}
