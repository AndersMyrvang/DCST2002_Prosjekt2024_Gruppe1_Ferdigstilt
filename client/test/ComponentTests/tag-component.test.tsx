import React from 'react';
import { render, screen, fireEvent, waitFor, act } from '@testing-library/react';
import { TagList, TagAdd } from '../../src/components/tag-component';
import tagService from '../../src/services/tag-service';
import userService from '../../src/services/user-service';
import { createHashHistory } from 'history';

jest.mock('../../src/services/tag-service');
jest.mock('../../src/services/user-service');
jest.mock('react-icons/fa', () => ({
  FaTrashAlt: () => <span data-testid="trash-icon">TrashIcon</span>,
}));
jest.mock('history', () => ({
  createHashHistory: jest.fn(() => ({
    push: jest.fn(),
  })),
}));

const history = createHashHistory();

describe('TagList Component', () => {
  const mockTags = [
    { tag_id: 1, tag_name: 'Tag1' },
    { tag_id: 2, tag_name: 'Tag2' },
  ];

  const mockUser = {
    user_id: 1,
    google_id: '123',
    username: 'testuser',
    email: 'test@test.com',
    profile_image_url: '',
    first_login: new Date(),
    last_login: new Date(),
    is_admin: true,
  };

  beforeEach(() => {
    jest.clearAllMocks();

    (tagService.getTags as jest.Mock).mockResolvedValue(mockTags);
    (tagService.getTagPageCount as jest.Mock).mockResolvedValue(0);
    (userService.getUserById as jest.Mock).mockResolvedValue(mockUser);

    localStorage.setItem('user_id', '1');
  });

  afterEach(() => {
    localStorage.removeItem('user_id');
  });

  test('renders tags and page counts', async () => {
    (tagService.getTagPageCount as jest.Mock).mockResolvedValueOnce(3).mockResolvedValueOnce(5);

    await act(async () => {
      render(<TagList />);
    });

    await waitFor(() => {
      expect(screen.getByText('Tag1')).toBeInTheDocument();
      expect(screen.getByText('Tag2')).toBeInTheDocument();
      expect(screen.getByText('Pages: 3')).toBeInTheDocument();
      expect(screen.getByText('Pages: 5')).toBeInTheDocument();
    });
  });

  test('allows admin to delete a tag', async () => {
    (tagService.deleteTag as jest.Mock).mockResolvedValue({});

    await act(async () => {
      render(<TagList />);
    });

    await waitFor(() => {
      expect(screen.getByText('Tag1')).toBeInTheDocument();
    });

    const deleteButton = screen.getAllByTestId('trash-icon')[0];
    window.confirm = jest.fn().mockReturnValueOnce(true);

    fireEvent.click(deleteButton);

    await waitFor(() => {
      expect(tagService.deleteTag).toHaveBeenCalledWith(1);
    });
  });

  test('hides delete button for non-admin users', async () => {
    const nonAdminUser = { ...mockUser, is_admin: false };
    (userService.getUserById as jest.Mock).mockResolvedValue(nonAdminUser);

    await act(async () => {
      render(<TagList />);
    });

    await waitFor(() => {
      expect(screen.queryByTestId('trash-icon')).not.toBeInTheDocument();
    });
  });

  test('renders "New tag" button for logged-in users', async () => {
    await act(async () => {
      render(<TagList />);
    });

    await waitFor(() => {
      expect(screen.getByText('New tag')).toBeInTheDocument();
    });
  });
});

describe('TagAdd Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  afterEach(() => {
    localStorage.removeItem('user_id');
  });

  test('matches snapshot for initial render', async () => {
    let container: DocumentFragment = document.createDocumentFragment();

    await act(async () => {
      const { asFragment } = render(<TagAdd />);
      container = asFragment();
    });

    expect(container).toMatchSnapshot();
  });

  test('adds a new tag', async () => {
    const mockUser = {
      user_id: 1,
      google_id: '123',
      username: 'testuser',
      email: 'test@test.no',
      profile_image_url: '',
      first_login: new Date(),
      last_login: new Date(),
      is_admin: true,
    };

    (userService.getUserById as jest.Mock).mockResolvedValue(mockUser);
    (tagService.addTag as jest.Mock).mockResolvedValue({ tag_id: 3, tag_name: 'New Tag' });

    await act(async () => {
      render(<TagAdd />);
    });

    const input = screen.getByPlaceholderText('Tag name');

    const button = screen.getByRole('button', { name: 'Add tag' });

    fireEvent.change(input, { target: { value: 'New Tag' } });
    fireEvent.click(button);

    await waitFor(() => {
      expect(tagService.addTag).toHaveBeenCalledWith({ tag_id: 0, tag_name: 'New Tag' });
    });
  });

  test('hides delete button for non-admin users', async () => {
    const nonAdminUser = {
      user_id: 10,
      google_id: '1234',
      username: 'testuser1',
      email: 'test1@test.no',
      profile_image_url: '',
      first_login: new Date(),
      last_login: new Date(),
      is_admin: false,
    };
    (userService.getUserById as jest.Mock).mockResolvedValue(nonAdminUser);

    await act(async () => {
      render(<TagAdd />);
    });

    await waitFor(() => {
      expect(screen.queryByTestId('trash-icon')).not.toBeInTheDocument();
    });
  });

  test('matches snapshot for initial render', async () => {
    let container: DocumentFragment = document.createDocumentFragment();

    await act(async () => {
      const { asFragment } = render(<TagAdd />);
      container = asFragment();
    });

    expect(container).toMatchSnapshot();
  });
});
