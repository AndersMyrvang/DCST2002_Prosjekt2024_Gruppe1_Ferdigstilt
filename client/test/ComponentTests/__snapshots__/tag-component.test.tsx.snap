// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagAdd Component matches snapshot for initial render 1`] = `<DocumentFragment />`;

exports[`TagAdd Component matches snapshot for initial render 2`] = `<DocumentFragment />`;
